### PR TITLE
Update RSVP and other definitions which depend on it with default export.

### PR DESCRIPTION
**Pull request by chriskrycho**
**Note:** To prevent a merge conflict in the Ember typings, I based this on #18379.

Docs for RSVP live in [its README] and the [Ember API docs].

[its README]: https://github.com/tildeio/rsvp.js/blob/81af1f409050ca3a05ae0aecafc89cd616f3bb13/README.md
[Ember API docs]: https://emberjs.com/api/ember/2.14/classes/RSVP/methods

The dual-authorship here is because @theroncross and I collaborated on these, and he's on vacation at present.

### CODE REVIEW

**Comment by chriskrycho** [comment]
Note: I'll push an update for the `ember-testing-helpers` issue this evening.

**Comment by chriskrycho** [comment]
e567181 is the last commit this requires. Feel free to review and merge at will!

**Comment by chriskrycho** [comment]
40d75ff is the last commit this requires. Feel free to review and merge at will!

**Comment by chriskrycho** [comment]
Found a bug (in `RSVP.hash()`), fixed the bug.

**Comment by mkohlmyr** [comment]
As far as RSVP goes I'm a big fan of the added tests, the new definitions look good to me but I don't have much time to look into it at the moment. That said if they pass the newly minted tests then that's good enough for me.

Can't comment on the ember stuff.

**Comment by chriskrycho** [comment]
For context there, the Ember bits (other than the commit I landed this on top of to prevent merge conflicts) are all just updates to reflect the fact that Ember just re-exports RSVP directly.

**Review by mkohlmyr — APPROVED** [review]
:+1: from me

**Comment by sheetalkamat** [comment]
@chriskrycho  can you please resolve the conflicts and update the PR. Thanks.

**Comment by chriskrycho** [comment]
@sheetalkamat done, and turned it into a nicely squashed and very descriptive single commit to merge in as well. Thanks!

**Comment by sheetalkamat** [comment]
@chriskrycho  In future please do not squash changes in the PR. It gets hard to verify if the changes are still correct. Merging of the PR will squash the changes.

**Comment by chriskrycho** [comment]
Ah, good to know! Sorry about that!

**Comment by tomdale** [comment]
@chriskrycho Some of these changes don't seem quite right to me.

At first glance:

1. The definition for `resolve` has regressed. Previously, `Promise.resolve` correctly expressed the assimilation behavior (`Promise.resolve<T>(object: Thenable<T>): T`). With these changes, promises will continue to nest (`Promise.resolve(Promise.resolve(true))` yields `Promise<Promise<boolean>>` instead of `Promise<boolean>`).
2. The requirement for the second generic type for Promises to capture the error type is a pretty big change. Is there a reason this deviates from TypeScript's built-in `Promise` types, and if so, is there a reason the error type doesn't default to `Error` or `never` to ease adoption?

**Comment by tomdale** [comment]
@chriskrycho Another problem is that `Promise.resolve()` is invalid with these changes, even though it is part of the specification (should return a promise that resolves to `undefined`). For compatibility reasons it really seems like we should be aligning with the built-in TypeScript Promise types and only deviating where necessary.

**Comment by chriskrycho** [comment]
@tomdale, thanks for bringing these up. I'm going to tackle them slightly out of sequence, with the most important bits at the end.

> 1. The definition for resolve has regressed. Previously, Promise.resolve correctly expressed the assimilation behavior (Promise.resolve<T>(object: Thenable<T>): T). With these changes, promises will continue to nest (Promise.resolve(Promise.resolve(true)) yields Promise<Promise<boolean>> instead of Promise<boolean>).

This is just a bug. The existing tests didn't flag it up; we just need to write a test that captures it so someone can do the type shenaniganry to make it fall through correctly. It's not something I usually write, so I forgot to check for it when reviewing the changes Theron made or when making further changes myself.

> Another problem is that `Promise.resolve()` is invalid with these changes, even though it is part of the specification (should return a promise that resolves to undefined). 

Likewise just something we missed, and easy to resolve. I think this comes out of the fact that TS doesn't allow `void` to inhabit generic types – at least, that seems to be the case in my testing; I haven't found that concretely documented anywhere. I only discovered that *after* tweaking these. We just need (a) a test for that and (b) an overload for that variant.

> 2. The requirement for the second generic type for Promises to capture the error type is a pretty big change. Is… is there a reason the error type doesn't default to `Error` or `never` to ease adoption?

I hadn't gotten my head around the various implications of default type parameters when I wrote it. I *think* this would do the trick for e.g. `catch`, though?

```typescript
catch<D = Error>(onRejection: Rejection<T, C, D>, label?: string): Promise<T, D>;
```

I don't *think* we want to default to `Promise<T, C = never>` for the default scenario – `Promise<T, C = Error>` seems like a better default?

> Is there a reason this deviates from TypeScript's built-in `Promise` types? [and] For compatibility reasons it really seems like we should be aligning with the built-in TypeScript Promise types and only deviating where necessary.

Basically, I was trying to deal with the fact that the TS definition leaves you completely blind about the  type of the error you're going to be handling, but I often both know that and want to be able to inform the consumer of an API (also often myself but a month or two later) about it. Without the second type parameter, it's always just `any`, which is *really* frustrating.

I'd be really happy to find a way to hew closer to the main TS definitions for Promises, but I also would like those definitions to be a lot more capable. Suggestions welcome, though.

**Comment by chriskrycho** [comment]
Oh, one other note: one of the problems is that TS doesn't let you specify *only one* type parameter, even if you have defaults for the others. It's quite *unlike* ES optional parameters that way. 😞 So if you type `RSVP.Promise<string>`, it'll complain that you didn't full specify the type, even if you defined it as `Promise<T, C = Error>`. Which I *also* didn't know at the time I wrote these, but which makes me feel like 🤦‍♂️.

**Comment by tomdale** [comment]
@chriskrycho Awesome, thank you! After looking at the test suite I realize it's somewhat anemic. I'm wondering if there's a more robust test suite (either for built-in Promises or bluebird?) that we might be able to borrow from.

> Basically, I was trying to deal with the fact that the TS definition leaves you completely blind about the type of the error you're going to be handling, but I often both know that and want to be able to inform the consumer of an API (also often myself but a month or two later) about it. Without the second type parameter, it's always just any, which is really frustrating.

That's fair, but it's possible it might lull you into a false sense of security, too. Why? Because of the default try/catch behavior of promises. Unlike `return` statements, which are relatively easy to reason about, exceptions, and thus promise rejections, can be generated _anywhere_ downstream. Unless you're doing something like `Promise.reject(someValue)`, you'd better be *darn sure* you're programming defensively against any possible error value caused by an exception in your code or in library code you're relying on.

**Comment by chriskrycho** [comment]
@tomdale that's a really fair point, and as I was thinking about it more over the last bit, I concluded roughly the same. It really does suck for one really specific condition, though:

```typescript
RSVP.reject("this is totally a string")
  .catch(reason => reason.length)
                       // ^ type error b/c reason is `any`, not `string`
```

**Comment by tomdale** [comment]
> It's quite unlike ES optional parameters that way. 😞 So if you type `RSVP.Promise<string>`, it'll complain that you didn't full specify the type, even if you defined it as `Promise<T, C = Error>`.

Are you sure? That's not the behavior I seem to be getting in TypeScript 2.4.2, but maybe I'm doing something different?

**Comment by tomdale** [comment]
@chriskrycho This is why I don't `throw` or `reject()` anything other than `Error` subclasses, for similar reasons. 😁

**Comment by chriskrycho** [comment]
> Are you sure? That's not the behavior I seem to be getting in TypeScript 2.4.2, but maybe I'm doing something different?

Oh, that's interesting; confirm – it *does* let you treat it the way it should. So that's something. I wonder if we could write it so that `RSVP.Promise<T, E = any>` extended native `PromiseLike<T>`. 🤔

> This is why I don't throw or reject() anything other than Error subclasses, for similar reasons. 😁

👏

**Comment by chriskrycho** [comment]
@tomdale oh, and yes _please_ to finding a more rigorous test suite.

**Comment by dwickern** [comment]
Maybe we can leverage the built-in interfaces `PromiseLike<T>` and `PromiseConstructor`.

This should be enough 
```
```
Typescript has some tests we could use, for example:
https://github.com/Microsoft/TypeScript/blob/master/tests/cases/compiler/promiseType.ts

**Comment by stefanpenner** [comment]
maintainer of RSVP here, I'm totally on-board with these definitions getting added to RSVP itself. Would welcome a PR.

**Comment by chriskrycho** [comment]
@tomdale @stefanpenner an update for you: I've *completely* reworked these typings [here](https://github.com/typed-ember/ember-typings/pull/42) and expect to be submitting a PR with them to RSVP itself after giving them a test-drive in my own Ember app on Monday.
